File: frontend/src/api/twoFactor.ts
/**
 * Two-Factor Authentication (TOTP) API client.
 *
 * All calls require the user to be authenticated (Bearer token sent by the
 * Axios interceptor in client.ts).  The sole exception is `login2FA`, which
 * accepts the short-lived temp_token returned by the normal login endpoint
 * when 2FA is required.
 */
import { api } from './client';

export interface TwoFAStatus {
  is_enabled: boolean;
  has_device: boolean;
}

export interface TwoFASetupResponse {
  secret: string;
  qr_code_url: string;
  backup_codes: string[];
}

export interface TwoFALoginResponse {
  access_token: string;
  refresh_token?: string;
  token_type: string;
  onboarding_completed?: boolean;
}

export interface BackupCodesResponse {
  backup_codes: string[];
  used_count: number;
  total: number;
}

export interface RegenerateBackupCodesResponse {
  detail: string;
  backup_codes: string[];
}

/** Check whether 2FA is enabled for the current user. */
export async function get2FAStatus(): Promise<TwoFAStatus> {
  const { data } = await api.get<TwoFAStatus>('/api/2fa/status');
  return data;
}

/** Initiate 2FA setup — returns secret, QR code, and backup codes. */
export async function setup2FA(): Promise<TwoFASetupResponse> {
  const { data } = await api.post<TwoFASetupResponse>('/api/2fa/setup');
  return data;
}

/** Verify a TOTP code and activate 2FA. */
export async function enable2FA(code: string): Promise<{ detail: string }> {
  const { data } = await api.post<{ detail: string }>('/api/2fa/enable', { code });
  return data;
}

/**
 * Verify a TOTP (or backup) code and deactivate 2FA.
 *
 * A valid code is required to prevent accidental / unauthorised disabling.
 */
export async function disable2FA(code: string): Promise<{ detail: string }> {
  const { data } = await api.post<{ detail: string }>('/api/2fa/disable', { code });
  return data;
}

/** Standalone TOTP / backup-code verification (not part of the login flow). */
export async function verify2FA(
  code: string,
): Promise<{ detail: string; backup_code_used: boolean }> {
  const { data } = await api.post<{ detail: string; backup_code_used: boolean }>(
    '/api/2fa/verify',
    { code },
  );
  return data;
}

/** Return masked backup codes for the current user. */
export async function getBackupCodes(): Promise<BackupCodesResponse> {
  const { data } = await api.get<BackupCodesResponse>('/api/2fa/backup-codes');
  return data;
}

/**
 * Regenerate backup codes (requires a valid TOTP code to authorise).
 *
 * Old codes are invalidated immediately.
 */
export async function regenerateBackupCodes(
  code: string,
): Promise<RegenerateBackupCodesResponse> {
  const { data } = await api.post<RegenerateBackupCodesResponse>(
    '/api/2fa/backup-codes/regenerate',
    { code },
  );
  return data;
}

/**
 * Complete the 2FA login flow.
 *
 * Called after POST /api/auth/login returns `{ requires_2fa: true, temp_token }`.
 * Accepts the temp token and the user's TOTP code (or backup code) and returns
 * a full JWT pair on success.
 */
export async function login2FA(
  tempToken: string,
  code: string,
): Promise<TwoFALoginResponse> {
  const { data } = await api.post<TwoFALoginResponse>('/api/auth/login/2fa', {
    temp_token: tempToken,
    code,
  });
  return data;
}

File: app/api/routes/two_factor.py
"""Two-factor authentication (TOTP) API routes.

Endpoints
---------
GET  /2fa/status                    — check if 2FA is enabled
POST /2fa/setup                     — generate secret + QR code
POST /2fa/enable                    — verify code and activate 2FA
POST /2fa/disable                   — verify code and deactivate 2FA
POST /2fa/verify                    — verify TOTP code (standalone, e.g. elevated actions)
GET  /2fa/backup-codes              — list masked backup codes
POST /2fa/backup-codes/regenerate   — regenerate backup codes
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.two_factor import TOTPDevice
from app.models.user import User
from app.schemas.two_factor import (
    BackupCodesResponse,
    TOTPSetupResponse,
    TOTPStatusResponse,
    TOTPVerifyRequest,
)
from app.services.two_factor import TwoFactorService

router = APIRouter(prefix="/2fa", tags=["Two-Factor Authentication"])
_logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Status
# ---------------------------------------------------------------------------

@router.get("/status", response_model=TOTPStatusResponse)
def get_2fa_status(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Return whether 2FA is set up and/or enabled for the current user."""
    device = (
        db.query(TOTPDevice)
        .filter(TOTPDevice.user_id == current_user.id)
        .first()
    )
    return TOTPStatusResponse(
        is_enabled=bool(device and device.is_enabled),
        has_device=device is not None,
    )


# ---------------------------------------------------------------------------
# Setup — generate secret + QR code
# ---------------------------------------------------------------------------

@router.post("/setup", response_model=TOTPSetupResponse)
def setup_2fa(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Initiate 2FA enrollment.

    Creates (or resets) a TOTP device for the current user and returns the
    secret, a QR code, and a fresh set of backup codes.  2FA is NOT yet
    enabled — the user must call POST /2fa/enable with a valid code first.
    """
    try:
        device, uri = TwoFactorService.create_totp_device(
            user_id=current_user.id,
            user_email=current_user.email or current_user.full_name,
            db=db,
        )
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        )

    try:
        qr_code_url = TwoFactorService.generate_qr_code_base64(uri)
    except RuntimeError as exc:
        # qrcode not installed — return the URI as a fallback
        _logger.warning("QR code generation unavailable: %s", exc)
        qr_code_url = uri  # frontend can render it with a JS library

    plaintext_secret = TwoFactorService.get_device_secret(device)
    return TOTPSetupResponse(
        secret=plaintext_secret,
        qr_code_url=qr_code_url,
        backup_codes=device.backup_codes or [],
    )


# ---------------------------------------------------------------------------
# Enable
# ---------------------------------------------------------------------------

@router.post("/enable")
def enable_2fa(
    body: TOTPVerifyRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Verify a TOTP code and activate 2FA for the current user."""
    device = (
        db.query(TOTPDevice)
        .filter(TOTPDevice.user_id == current_user.id)
        .first()
    )
    if not device:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No TOTP device found. Call POST /2fa/setup first.",
        )
    if device.is_enabled:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="2FA is already enabled.",
        )

    try:
        secret = TwoFactorService.get_device_secret(device)
        valid = TwoFactorService.verify_totp(secret, body.code)
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        )

    if not valid:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid TOTP code.",
        )

    TwoFactorService.enable_device(device, db)
    return {"detail": "Two-factor authentication enabled successfully."}


# ---------------------------------------------------------------------------
# Disable
# ---------------------------------------------------------------------------

@router.post("/disable")
def disable_2fa(
    body: TOTPVerifyRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Verify a TOTP code (or backup code) and deactivate 2FA."""
    device = (
        db.query(TOTPDevice)
        .filter(TOTPDevice.user_id == current_user.id)
        .first()
    )
    if not device:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="2FA is not configured for this account.",
        )
    if not device.is_enabled:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="2FA is already disabled.",
        )

    # Accept a valid TOTP code OR an unused backup code
    code_ok = False
    try:
        secret = TwoFactorService.get_device_secret(device)
        code_ok = TwoFactorService.verify_totp(secret, body.code)
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        )

    if not code_ok:
        code_ok = TwoFactorService.verify_backup_code(device, body.code)

    if not code_ok:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid code.",
        )

    TwoFactorService.disable_device(device, db)
    return {"detail": "Two-factor authentication disabled."}


# ---------------------------------------------------------------------------
# Verify (standalone — for elevated actions, not login)
# ---------------------------------------------------------------------------

@router.post("/verify")
def verify_2fa(
    body: TOTPVerifyRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Verify a TOTP code or backup code for the authenticated user.

    This is a general-purpose verification endpoint (e.g. before a sensitive
    action).  It is NOT part of the login flow — see POST /auth/login/2fa.
    """
    device = (
        db.query(TOTPDevice)
        .filter(TOTPDevice.user_id == current_user.id, TOTPDevice.is_enabled == True)  # noqa: E712
        .first()
    )
    if not device:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="2FA is not enabled for this account.",
        )

    try:
        secret = TwoFactorService.get_device_secret(device)
        code_ok = TwoFactorService.verify_totp(secret, body.code)
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        )

    if not code_ok:
        # Try backup code
        if TwoFactorService.verify_backup_code(device, body.code):
            TwoFactorService.use_backup_code(device, body.code, db)
            return {"detail": "Backup code accepted.", "backup_code_used": True}
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid TOTP code.",
        )

    return {"detail": "Code verified.", "backup_code_used": False}


# ---------------------------------------------------------------------------
# Backup codes
# ---------------------------------------------------------------------------

@router.get("/backup-codes", response_model=BackupCodesResponse)
def get_backup_codes(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Return masked backup codes (first 2 chars visible, rest hidden)."""
    device = (
        db.query(TOTPDevice)
        .filter(TOTPDevice.user_id == current_user.id)
        .first()
    )
    if not device:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="2FA is not configured for this account.",
        )

    masked = TwoFactorService.mask_backup_codes(
        device.backup_codes or [], device.used_backup_codes or []
    )
    return BackupCodesResponse(
        backup_codes=masked,
        used_count=len(device.used_backup_codes or []),
        total=len(device.backup_codes or []),
    )


@router.post("/backup-codes/regenerate")
def regenerate_backup_codes(
    body: TOTPVerifyRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Verify a TOTP code and generate a fresh set of backup codes.

    Old backup codes are invalidated immediately.
    """
    device = (
        db.query(TOTPDevice)
        .filter(TOTPDevice.user_id == current_user.id, TOTPDevice.is_enabled == True)  # noqa: E712
        .first()
    )
    if not device:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="2FA is not enabled for this account.",
        )

    try:
        secret = TwoFactorService.get_device_secret(device)
        valid = TwoFactorService.verify_totp(secret, body.code)
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        )

    if not valid:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid TOTP code.",
        )

    new_codes = TwoFactorService.regenerate_backup_codes(device, db)
    return {
        "detail": "Backup codes regenerated. Save them somewhere safe!",
        "backup_codes": new_codes,
    }
